peep.c

```diff
@@ -3868,6 +3868,98 @@ Perl_rpeep(pTHX_ OP *o)
             }
             break;
 
+        case OP_SUBSTR: {
+            OP *expr, *offs, *len;
+            /* Specialize substr($x, 0, $y) and substr($x,0,$y,"") */
+            /* Does this substr have 3-4 args and amiable flags? */
+            if (
+                ((cMAXARG3x(o) == 4) || (cMAXARG3x(o) == 3))
+                /* No lvalue cases, no OPpSUBSTR_REPL_FIRST*/
+                && !(o->op_private & (OPpSUBSTR_REPL_FIRST|OPpMAYBE_LVSUB))
+                && !(o->op_flags & OPf_MOD)
+            ){
+                /* Should be a leading ex-pushmark */
+                OP *pushmark = cBINOPx(o)->op_first;
+                assert(pushmark->op_type == OP_NULL);
+                expr = OpSIBLING(pushmark);
+                offs = OpSIBLING(expr);
+
+                /* Gets complicated fast if the expr isn't simple*/
+                if (expr->op_type != OP_PADSV)
+                    break;
+                /* Is the offset CONST zero? */
+                if (offs->op_type != OP_CONST)
+                    break;
+                SV *offs_sv = cSVOPx_sv(offs);
+                if (!(SvIOK(offs_sv) && SvIVX(offs_sv) == 0))
+                    break;
+                len  = OpSIBLING(offs);
+
+                if (cMAXARG3x(o) == 4) {/* replacement */
+                    /* Is the replacement string CONST ""? */
+                    OP *repl = OpSIBLING(len);
+                    if (repl->op_type != OP_CONST)
+                        break;
+                    SV *repl_sv = cSVOPx_sv(repl);
+                    if(!(SvPOK(repl_sv) && SvCUR(repl_sv) == 0))
+                        break;
+                }
+            } else {
+                break;
+            }
+            /* It's on! */
+            /* Take out the static LENGTH & REPLACMENT OPs */
+            /* (The finalizer does not seem to change op_next here) */
+            expr->op_next = offs->op_next;
+            o->op_private = cMAXARG3x(o);
+            if (cMAXARG3x(o) == 4)
+                len->op_next = o;
+
+            /* We have a problem if padrange pushes EXRP for us and
```

Comment:
"EXPR"?

Reply:
As in "substr EXPR,OFFSET,LENGTH,REPLACEMENT". I'll change that to something more legible!

Reply:
Oh yeah, typo too. :sob:

Reply:
Yeah, it was about the typo :)